File: .claude/skills/ExpandedPasswordList/Tools/gravel_filter.py
#!/usr/bin/env python3
"""
gravel_filter.py - ROCKS -> GRAVEL Filter (1:1 batch correspondence)

Reads rockyou.txt, computes SHA-1 hashes in memory, filters each ROCKS batch
to produce a corresponding GRAVEL batch. Also saves hash:plaintext pairs for
every rockyou match into rocks/batch-NNNN-pairs.txt.

Invariant: rocks/batch-NNNN.txt - SHA1(rockyou.txt) = gravel/batch-NNNN.txt

No pre-computed indexes. No trust in intermediate files. Straight from source.

Output per batch:
  gravel/batch-NNNN.txt        Uncracked hashes (gravel)
  rocks/batch-NNNN-pairs.txt   Cracked hash:plaintext pairs

Memory: ~2-3GB (14.3M SHA-1 -> plaintext dict)

Usage:
  python gravel_filter.py                    Filter all (with resume)
  python gravel_filter.py --no-resume        Start fresh
  python gravel_filter.py --verify           Verify invariant
  python gravel_filter.py --status           Show progress
  python gravel_filter.py --rockyou PATH     Alternate rockyou.txt
"""

import hashlib
import json
import os
import re
import sys
import time
from pathlib import Path

# Resolve paths
SCRIPT_DIR = Path(__file__).parent
SKILL_DIR = SCRIPT_DIR.parent
PROJECT_ROOT = SKILL_DIR.parent.parent.parent

# Data dir: check for symlink/directory at skill level
DATA_DIR = SKILL_DIR / "data"
if not DATA_DIR.exists():
    print(f"ERROR: Data directory not found: {DATA_DIR}")
    sys.exit(1)

ROCKS_DIR = DATA_DIR / "rocks"
GRAVEL_DIR = DATA_DIR / "gravel"
DEFAULT_ROCKYOU = PROJECT_ROOT / "rockyou.txt"
PROGRESS_FILE = DATA_DIR / "gravel-filter-progress.json"


def fmt_duration(seconds):
    if seconds < 60:
        return f"{seconds:.0f}s"
    if seconds < 3600:
        return f"{int(seconds // 60)}m {int(seconds % 60)}s"
    hours = int(seconds // 3600)
    mins = int((seconds % 3600) // 60)
    return f"{hours}h {mins}m"


def fmt_num(n):
    return f"{n:,}"


def load_progress():
    if PROGRESS_FILE.exists():
        return json.loads(PROGRESS_FILE.read_text())
    return {
        "lastCompletedBatch": 0,
        "totalRocksHashes": 0,
        "totalRockyouFiltered": 0,
        "totalGravelHashes": 0,
        "batchesProcessed": 0,
        "rockyouEntries": 0,
        "rockyouPath": "",
        "startedAt": time.strftime("%Y-%m-%dT%H:%M:%S"),
        "lastUpdated": time.strftime("%Y-%m-%dT%H:%M:%S"),
    }


def save_progress(progress):
    progress["lastUpdated"] = time.strftime("%Y-%m-%dT%H:%M:%S")
    PROGRESS_FILE.write_text(json.dumps(progress, indent=2))


def build_rockyou_hash_dict(rockyou_path):
    """Read rockyou.txt, SHA-1 hash every entry, return as dict {uppercase_hex: plaintext}.

    For duplicate SHA-1s (different entries producing same hash), keeps the first entry.
    """
    print(f"Reading rockyou.txt from: {rockyou_path}")
    start = time.time()

    hash_dict = {}
    processed = 0
    skipped = 0

    with open(rockyou_path, "r", encoding="utf-8", errors="surrogateescape") as f:
        for line in f:
            entry = line.rstrip("\n").rstrip("\r")
            if not entry:
                skipped += 1
                continue

            sha1 = hashlib.sha1(entry.encode("utf-8", errors="surrogateescape")).hexdigest().upper()
            if sha1 not in hash_dict:
                hash_dict[sha1] = entry
            processed += 1

            if processed % 1_000_000 == 0:
                elapsed = time.time() - start
                print(f"    {processed // 1_000_000}M hashed... ({fmt_duration(elapsed)})")

    elapsed = time.time() - start
    dupes = processed - len(hash_dict)
    print(f"  Done: {fmt_num(len(hash_dict))} unique SHA-1 hashes from {fmt_num(processed)} entries")
    print(f"  Skipped: {fmt_num(skipped)} empty lines")
    print(f"  Duplicates: {fmt_num(dupes)}")
    print(f"  Time: {fmt_duration(elapsed)}")
    return hash_dict


def get_rocks_batches():
    """Get sorted list of rocks batch files."""
    if not ROCKS_DIR.exists():
        return []
    batches = []
    for f in ROCKS_DIR.iterdir():
        m = re.match(r"^batch-(\d+)\.txt$", f.name)
        if m:
            batches.append({"number": int(m.group(1)), "filename": f.name})
    return sorted(batches, key=lambda b: b["number"])


def filter_batch(rockyou_hash_dict, rocks_file, gravel_file, pairs_file):
    """Filter a single rocks batch -> gravel batch + pairs file.

    Streams through rocks file line by line, writing gravel and pairs
    directly to output files. No intermediate lists held in memory.

    Args:
        rockyou_hash_dict: dict {SHA1_HEX: plaintext}
        rocks_file: path to rocks/batch-NNNN.txt
        gravel_file: path to gravel/batch-NNNN.txt (uncracked output)
        pairs_file: path to rocks/batch-NNNN-pairs.txt (hash:plaintext output)
    """
    rocks_count = 0
    filtered_count = 0
    gravel_count = 0

    with open(rocks_file, "r", encoding="utf-8", errors="surrogateescape") as rf, \
         open(gravel_file, "w", encoding="utf-8") as gf, \
         open(pairs_file, "w", encoding="utf-8", errors="surrogateescape") as pf:
        for line in rf:
            h = line.strip()
            if not h:
                continue
            rocks_count += 1
            plaintext = rockyou_hash_dict.get(h.upper())
            if plaintext is not None:
                pf.write(f"{h.upper()}:{plaintext}\n")
                filtered_count += 1
            else:
                gf.write(h + "\n")
                gravel_count += 1

    return {"rocksCount": rocks_count, "filteredCount": filtered_count, "gravelCount": gravel_count}


def filter_all(resume=True, rockyou_path=None):
    rockyou_path = rockyou_path or DEFAULT_ROCKYOU
    if not os.path.exists(rockyou_path):
        print(f"ERROR: rockyou.txt not found: {rockyou_path}")
        sys.exit(1)

    GRAVEL_DIR.mkdir(parents=True, exist_ok=True)

    rocks_batches = get_rocks_batches()
    if not rocks_batches:
        print(f"ERROR: No rocks batches in {ROCKS_DIR}")
        print("Run RocksExtractor first")
        sys.exit(1)

    # Build hash dict from rockyou.txt (SHA1 -> plaintext)
    rockyou_hashes = build_rockyou_hash_dict(rockyou_path)

    if resume:
        progress = load_progress()
    else:
        progress = load_progress()
        progress["lastCompletedBatch"] = 0
        progress["totalRocksHashes"] = 0
        progress["totalRockyouFiltered"] = 0
        progress["totalGravelHashes"] = 0
        progress["batchesProcessed"] = 0
        progress["startedAt"] = time.strftime("%Y-%m-%dT%H:%M:%S")

    progress["rockyouEntries"] = len(rockyou_hashes)
    progress["rockyouPath"] = str(rockyou_path)

    pending = [b for b in rocks_batches if b["number"] > progress["lastCompletedBatch"]]

    print()
    print("GravelFilter - ROCKS -> GRAVEL Filter")
    print("=====================================")
    print(f"Rockyou source: {rockyou_path}")
    print(f"Rockyou hashes: {fmt_num(len(rockyou_hashes))}")
    print(f"ROCKS source:   {ROCKS_DIR}")
    print(f"GRAVEL output:  {GRAVEL_DIR}")
    print(f"PAIRS output:   {ROCKS_DIR}/batch-NNNN-pairs.txt")
    print(f"Total batches:  {len(rocks_batches)}")
    print(f"Already done:   {progress['batchesProcessed']}")
    print(f"Pending:        {len(pending)}")
    print(f"Mode:           1:1 (rocks[N] - SHA1(rockyou) = gravel[N])")
    print()

    if not pending:
        print("All batches already filtered!")
        return

    start = time.time()

    for i, batch in enumerate(pending):
        batch_name = f"batch-{batch['number']:04d}"
        rocks_file = ROCKS_DIR / batch["filename"]
        gravel_file = GRAVEL_DIR / f"{batch_name}.txt"
        pairs_file = ROCKS_DIR / f"{batch_name}-pairs.txt"

        result = filter_batch(rockyou_hashes, rocks_file, gravel_file, pairs_file)

        progress["totalRocksHashes"] += result["rocksCount"]
        progress["totalRockyouFiltered"] += result["filteredCount"]
        progress["totalGravelHashes"] += result["gravelCount"]
        progress["batchesProcessed"] += 1
        progress["lastCompletedBatch"] = batch["number"]

        invariant_ok = result["rocksCount"] == result["filteredCount"] + result["gravelCount"]

        elapsed = time.time() - start
        done = i + 1
        rate = done / elapsed if elapsed > 0 else 0
        remaining = len(pending) - done
        eta = remaining / rate if rate > 0 else 0
        filter_pct = (result["filteredCount"] / result["rocksCount"] * 100) if result["rocksCount"] > 0 else 0

        status = "OK" if invariant_ok else "MISMATCH!"
        print(
            f"  {batch_name}: {fmt_num(result['rocksCount'])} rocks -> "
            f"{fmt_num(result['gravelCount'])} gravel ({filter_pct:.1f}% filtered) "
            f"{status} | {done}/{len(pending)} | ETA: {fmt_duration(eta)}"
        )

        if not invariant_ok:
            print(f"  INVARIANT VIOLATION: {result['rocksCount']} != {result['filteredCount']} + {result['gravelCount']}")

        if done % 50 == 0:
            save_progress(progress)

    save_progress(progress)

    total_time = time.time() - start
    overall_rate = (progress["totalRockyouFiltered"] / progress["totalRocksHashes"] * 100) if progress["totalRocksHashes"] > 0 else 0

    print()
    print("Filter Complete")
    print("===============")
    print(f"Batches processed:      {progress['batchesProcessed']}")
    print(f"Total ROCKS hashes:     {fmt_num(progress['totalRocksHashes'])}")
    print(f"Rockyou filtered:       {fmt_num(progress['totalRockyouFiltered'])} ({overall_rate:.2f}%)")
    print(f"GRAVEL output:          {fmt_num(progress['totalGravelHashes'])}")
    print(f"Pairs files:            {ROCKS_DIR}/batch-NNNN-pairs.txt")
    print(f"Time:                   {fmt_duration(total_time)}")
    print(f"Invariant:              rocks[N] - SHA1(rockyou) = gravel[N]")


def verify(rockyou_path=None):
    rockyou_path = rockyou_path or DEFAULT_ROCKYOU
    rockyou_hashes = build_rockyou_hash_dict(rockyou_path)
    rocks_batches = get_rocks_batches()
    violations = 0
    verified = 0

    print(f"\nVerifying {len(rocks_batches)} batch pairs...")

    for batch in rocks_batches:
        batch_name = f"batch-{batch['number']:04d}"
        rocks_file = ROCKS_DIR / batch["filename"]
        gravel_file = GRAVEL_DIR / f"{batch_name}.txt"

        if not gravel_file.exists():
            print(f"  {batch_name}: MISSING gravel file")
            violations += 1
            continue

        with open(rocks_file) as f:
            rocks_lines = [l.strip() for l in f if l.strip()]
        with open(gravel_file) as f:
            gravel_set = {l.strip() for l in f if l.strip()}

        rocks_set = set(rocks_lines)
        rockyou_in_gravel = sum(1 for h in gravel_set if h.upper() in rockyou_hashes)
        gravel_not_in_rocks = sum(1 for h in gravel_set if h not in rocks_set)
        unaccounted = sum(1 for h in rocks_lines if h not in gravel_set and h.upper() not in rockyou_hashes)

        if rockyou_in_gravel > 0 or gravel_not_in_rocks > 0 or unaccounted > 0:
            print(f"  {batch_name}: VIOLATION — rockyou_in_gravel={rockyou_in_gravel}, gravel_not_in_rocks={gravel_not_in_rocks}, unaccounted={unaccounted}")
            violations += 1
        else:
            verified += 1

        if (verified + violations) % 100 == 0:
            print(f"  ... {verified + violations}/{len(rocks_batches)} checked")

    print(f"\nVerification: {verified} OK, {violations} violations out of {len(rocks_batches)} batches")


def show_status():
    progress = load_progress()
    rocks_batches = get_rocks_batches()
    gravel_count = len([f for f in GRAVEL_DIR.iterdir() if f.name.startswith("batch-") and f.name.endswith(".txt")]) if GRAVEL_DIR.exists() else 0

    print()
    print("GravelFilter Status")
    print("===================")
    print(f"Rockyou source:    {progress.get('rockyouPath', '(not set)')}")
    print(f"Rockyou hashes:    {fmt_num(progress.get('rockyouEntries', 0))}")
    print(f"ROCKS batches:     {len(rocks_batches)}")
    print(f"GRAVEL batches:    {gravel_count}")
    print(f"Batches filtered:  {progress['batchesProcessed']}")
    print(f"ROCKS hashes:      {fmt_num(progress['totalRocksHashes'])}")
    print(f"Rockyou filtered:  {fmt_num(progress['totalRockyouFiltered'])}")
    print(f"GRAVEL hashes:     {fmt_num(progress['totalGravelHashes'])}")
    print(f"Started:           {progress['startedAt']}")
    print(f"Last updated:      {progress['lastUpdated']}")


if __name__ == "__main__":
    args = sys.argv[1:]

    rockyou_path = DEFAULT_ROCKYOU
    if "--rockyou" in args:
        idx = args.index("--rockyou")
        if idx + 1 < len(args):
            rockyou_path = Path(args[idx + 1])

    if "--help" in args or "-h" in args:
        print(__doc__)
        sys.exit(0)
    elif "--status" in args:
        show_status()
    elif "--verify" in args:
        verify(rockyou_path)
    elif "--no-resume" in args:
        filter_all(resume=False, rockyou_path=rockyou_path)
    else:
        filter_all(resume=True, rockyou_path=rockyou_path)

File: .claude/skills/ExpandedPasswordList/Tools/GravelFilter.ts
#!/usr/bin/env bun
/**
 * GravelFilter.ts - ROCKS → GRAVEL Filter (1:1 batch correspondence)
 *
 * Reads rockyou.txt from disk, computes SHA-1 hashes in memory, then filters
 * each ROCKS batch to produce a corresponding GRAVEL batch.
 *
 * Invariant: rocks/batch-NNNN.txt - SHA1(rockyou.txt) = gravel/batch-NNNN.txt
 *
 * No pre-computed indexes. No trust in intermediate files. Straight from source.
 *
 * Pipeline: rocks/ → [THIS] → gravel/
 *
 * Memory: ~3GB (14.3M SHA-1 hashes in Set). Use: bun --max-old-space-size=4096
 *
 * @author PAI (Personal AI Infrastructure)
 * @license MIT
 */

import { existsSync, mkdirSync, writeFileSync, readFileSync, readdirSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { createHash } from "node:crypto";
import { DATA_DIR, ROCKS_DIR, GRAVEL_DIR } from "./config";

// =============================================================================
// Paths
// =============================================================================

const CURRENT_FILE = fileURLToPath(import.meta.url);
const TOOLS_DIR = dirname(CURRENT_FILE);
const SKILL_DIR = dirname(TOOLS_DIR);
const PROJECT_ROOT = resolve(SKILL_DIR, "..", "..", "..");
const DEFAULT_ROCKYOU = resolve(PROJECT_ROOT, "rockyou.txt");
const PROGRESS_FILE = resolve(DATA_DIR, "gravel-filter-progress.json");

// =============================================================================
// Progress Tracking
// =============================================================================

interface FilterProgress {
  lastCompletedBatch: number;
  totalRocksHashes: number;
  totalRockyouFiltered: number;
  totalGravelHashes: number;
  batchesProcessed: number;
  rockyouEntries: number;
  rockyouPath: string;
  startedAt: string;
  lastUpdated: string;
}

function loadProgress(): FilterProgress {
  if (existsSync(PROGRESS_FILE)) {
    return JSON.parse(readFileSync(PROGRESS_FILE, "utf-8"));
  }
  return {
    lastCompletedBatch: 0,
    totalRocksHashes: 0,
    totalRockyouFiltered: 0,
    totalGravelHashes: 0,
    batchesProcessed: 0,
    rockyouEntries: 0,
    rockyouPath: "",
    startedAt: new Date().toISOString(),
    lastUpdated: new Date().toISOString(),
  };
}

function saveProgress(progress: FilterProgress): void {
  progress.lastUpdated = new Date().toISOString();
  writeFileSync(PROGRESS_FILE, JSON.stringify(progress, null, 2));
}

// =============================================================================
// Utilities
// =============================================================================

function formatDuration(seconds: number): string {
  if (seconds < 60) return `${seconds.toFixed(0)}s`;
  if (seconds < 3600) return `${Math.floor(seconds / 60)}m ${Math.floor(seconds % 60)}s`;
  const hours = Math.floor(seconds / 3600);
  const mins = Math.floor((seconds % 3600) / 60);
  return `${hours}h ${mins}m`;
}

function formatBytes(bytes: number): string {
  if (bytes < 1024) return `${bytes}B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)}KB`;
  if (bytes < 1024 * 1024 * 1024) return `${(bytes / (1024 * 1024)).toFixed(1)}MB`;
  return `${(bytes / (1024 * 1024 * 1024)).toFixed(2)}GB`;
}

// =============================================================================
// Core: Build rockyou SHA-1 Set from source
// =============================================================================

/**
 * Read rockyou.txt, SHA-1 hash every entry, return as Set<string> (uppercase hex).
 * This is the single source of truth — no pre-computed indexes.
 */
function buildRockyouHashSet(rockyouPath: string): Set<string> {
  console.log(`Reading rockyou.txt from: ${rockyouPath}`);
  const startTime = Date.now();

  const content = readFileSync(rockyouPath, "utf-8");
  const lines = content.split("\n");

  console.log(`  ${lines.length.toLocaleString()} lines read (${formatBytes(content.length)})`);
  console.log("  Computing SHA-1 hashes...");

  const hashSet = new Set<string>();
  let processed = 0;
  let skipped = 0;

  for (const line of lines) {
    // Skip empty lines
    if (line.length === 0) {
      skipped++;
      continue;
    }

    // Strip trailing \r if present (Windows line endings)
    const entry = line.endsWith("\r") ? line.slice(0, -1) : line;
    if (entry.length === 0) {
      skipped++;
      continue;
    }

    const sha1 = createHash("sha1").update(entry, "utf-8").digest("hex").toUpperCase();
    hashSet.add(sha1);
    processed++;

    // Progress every 1M entries
    if (processed % 1_000_000 === 0) {
      const elapsed = (Date.now() - startTime) / 1000;
      const heapMB = Math.round(process.memoryUsage().heapUsed / (1024 * 1024));
      console.log(`    ${(processed / 1_000_000).toFixed(0)}M hashed... (${heapMB}MB heap, ${formatDuration(elapsed)})`);
    }
  }

  const elapsed = (Date.now() - startTime) / 1000;
  const heapMB = Math.round(process.memoryUsage().heapUsed / (1024 * 1024));

  console.log(`  Done: ${hashSet.size.toLocaleString()} unique SHA-1 hashes from ${processed.toLocaleString()} entries`);
  console.log(`  Skipped: ${skipped.toLocaleString()} empty lines`);
  console.log(`  Duplicates: ${(processed - hashSet.size).toLocaleString()}`);
  console.log(`  Time: ${formatDuration(elapsed)} | Heap: ${heapMB}MB`);

  return hashSet;
}

// =============================================================================
// Batch Processing
// =============================================================================

function getRocksBatches(): { number: number; filename: string }[] {
  if (!existsSync(ROCKS_DIR)) return [];

  return readdirSync(ROCKS_DIR)
    .filter((f) => f.startsWith("batch-") && f.endsWith(".txt"))
    .map((f) => {
      const match = f.match(/^batch-(\d+)\.txt$/);
      return match ? { number: parseInt(match[1]), filename: f } : null;
    })
    .filter((x): x is { number: number; filename: string } => x !== null)
    .sort((a, b) => a.number - b.number);
}

/**
 * Filter a single rocks batch → gravel batch using the in-memory hash Set
 */
function filterBatch(
  rockyouHashes: Set<string>,
  rocksBatchFile: string,
  gravelBatchFile: string
): { rocksCount: number; filteredCount: number; gravelCount: number } {
  const content = readFileSync(rocksBatchFile, "utf-8");
  const lines = content.split("\n").filter((l) => l.length > 0);

  const gravelHashes: string[] = [];
  let filteredCount = 0;

  for (const hash of lines) {
    if (rockyouHashes.has(hash.toUpperCase())) {
      filteredCount++;
    } else {
      gravelHashes.push(hash);
    }
  }

  // Write gravel batch (even if empty — maintains 1:1 correspondence)
  writeFileSync(gravelBatchFile, gravelHashes.join("\n") + "\n");

  return {
    rocksCount: lines.length,
    filteredCount,
    gravelCount: gravelHashes.length,
  };
}

/**
 * Filter all rocks batches → gravel batches
 */
async function filterAllBatches(options: {
  resume?: boolean;
  rockyouPath?: string;
} = {}): Promise<void> {
  const { resume = true, rockyouPath = DEFAULT_ROCKYOU } = options;

  // Validate rockyou.txt exists
  if (!existsSync(rockyouPath)) {
    console.error(`rockyou.txt not found: ${rockyouPath}`);
    console.error("Provide path: bun Tools/GravelFilter.ts --rockyou /path/to/rockyou.txt");
    process.exit(1);
  }

  // Ensure gravel dir exists
  if (!existsSync(GRAVEL_DIR)) {
    mkdirSync(GRAVEL_DIR, { recursive: true });
  }

  // Get rocks batches
  const rocksBatches = getRocksBatches();
  if (rocksBatches.length === 0) {
    console.error(`No rocks batches found in ${ROCKS_DIR}`);
    console.error("Run: bun Tools/RocksExtractor.ts first");
    process.exit(1);
  }

  // Build SHA-1 hash set from rockyou.txt (the expensive part)
  const rockyouHashes = buildRockyouHashSet(rockyouPath);

  let progress: FilterProgress;
  if (resume) {
    progress = loadProgress();
    // If resuming but rockyou path changed, warn
    if (progress.rockyouPath && progress.rockyouPath !== rockyouPath) {
      console.warn(`\nWARNING: rockyou path changed since last run!`);
      console.warn(`  Previous: ${progress.rockyouPath}`);
      console.warn(`  Current:  ${rockyouPath}`);
      console.warn(`  Use --no-resume to start fresh\n`);
    }
  } else {
    progress = {
      lastCompletedBatch: 0,
      totalRocksHashes: 0,
      totalRockyouFiltered: 0,
      totalGravelHashes: 0,
      batchesProcessed: 0,
      rockyouEntries: rockyouHashes.size,
      rockyouPath,
      startedAt: new Date().toISOString(),
      lastUpdated: new Date().toISOString(),
    };
  }

  // Update rockyou metadata
  progress.rockyouEntries = rockyouHashes.size;
  progress.rockyouPath = rockyouPath;

  // Filter to pending batches
  const pendingBatches = rocksBatches.filter((b) => b.number > progress.lastCompletedBatch);

  console.log("");
  console.log("GravelFilter - ROCKS → GRAVEL Filter");
  console.log("============================================");
  console.log(`Rockyou source: ${rockyouPath}`);
  console.log(`Rockyou hashes: ${rockyouHashes.size.toLocaleString()}`);
  console.log(`ROCKS source:   ${ROCKS_DIR}`);
  console.log(`GRAVEL output:  ${GRAVEL_DIR}`);
  console.log(`Total batches:  ${rocksBatches.length}`);
  console.log(`Already done:   ${progress.batchesProcessed}`);
  console.log(`Pending:        ${pendingBatches.length}`);
  console.log(`Mode:           1:1 (rocks[N] - SHA1(rockyou) = gravel[N])`);
  console.log("");

  if (pendingBatches.length === 0) {
    console.log("All batches already filtered!");
    return;
  }

  const startTime = Date.now();

  for (let i = 0; i < pendingBatches.length; i++) {
    const batch = pendingBatches[i];
    const batchName = `batch-${String(batch.number).padStart(4, "0")}`;
    const rocksFile = resolve(ROCKS_DIR, batch.filename);
    const gravelFile = resolve(GRAVEL_DIR, `${batchName}.txt`);

    const result = filterBatch(rockyouHashes, rocksFile, gravelFile);

    progress.totalRocksHashes += result.rocksCount;
    progress.totalRockyouFiltered += result.filteredCount;
    progress.totalGravelHashes += result.gravelCount;
    progress.batchesProcessed++;
    progress.lastCompletedBatch = batch.number;

    // Verify invariant
    const invariantOk = result.rocksCount === result.filteredCount + result.gravelCount;

    // Progress report
    const elapsed = (Date.now() - startTime) / 1000;
    const processed = i + 1;
    const rate = processed / elapsed;
    const remaining = pendingBatches.length - processed;
    const eta = rate > 0 ? remaining / rate : 0;
    const filterPct = result.rocksCount > 0
      ? ((result.filteredCount / result.rocksCount) * 100).toFixed(1)
      : "0";

    console.log(
      `  ${batchName}: ${result.rocksCount.toLocaleString()} rocks → ` +
        `${result.gravelCount.toLocaleString()} gravel (${filterPct}% filtered) ` +
        `${invariantOk ? "OK" : "MISMATCH!"} | ` +
        `${processed}/${pendingBatches.length} | ` +
        `ETA: ${formatDuration(eta)}`
    );

    if (!invariantOk) {
      console.error(`  INVARIANT VIOLATION: ${result.rocksCount} != ${result.filteredCount} + ${result.gravelCount}`);
    }

    // Save progress every 50 batches
    if (processed % 50 === 0) {
      saveProgress(progress);
    }
  }

  // Final save
  saveProgress(progress);

  const totalTime = (Date.now() - startTime) / 1000;
  const overallFilterRate = progress.totalRocksHashes > 0
    ? ((progress.totalRockyouFiltered / progress.totalRocksHashes) * 100).toFixed(2)
    : "0";

  console.log("");
  console.log("Filter Complete");
  console.log("===============");
  console.log(`Batches processed:      ${progress.batchesProcessed}`);
  console.log(`Total ROCKS hashes:     ${progress.totalRocksHashes.toLocaleString()}`);
  console.log(`Rockyou filtered:       ${progress.totalRockyouFiltered.toLocaleString()} (${overallFilterRate}%)`);
  console.log(`GRAVEL output:          ${progress.totalGravelHashes.toLocaleString()}`);
  console.log(`Time:                   ${formatDuration(totalTime)}`);
  console.log(`Invariant:              rocks[N] - SHA1(rockyou) = gravel[N]`);
}

/**
 * Verify existing rocks/gravel 1:1 correspondence
 */
async function verifyBatches(rockyouPath: string): Promise<void> {
  if (!existsSync(rockyouPath)) {
    console.error(`rockyou.txt not found: ${rockyouPath}`);
    process.exit(1);
  }

  const rockyouHashes = buildRockyouHashSet(rockyouPath);
  const rocksBatches = getRocksBatches();
  let violations = 0;
  let verified = 0;

  console.log(`\nVerifying ${rocksBatches.length} batch pairs...`);

  for (const batch of rocksBatches) {
    const batchName = `batch-${String(batch.number).padStart(4, "0")}`;
    const rocksFile = resolve(ROCKS_DIR, batch.filename);
    const gravelFile = resolve(GRAVEL_DIR, `${batchName}.txt`);

    if (!existsSync(gravelFile)) {
      console.log(`  ${batchName}: MISSING gravel file`);
      violations++;
      continue;
    }

    const rocksLines = readFileSync(rocksFile, "utf-8").split("\n").filter((l) => l.length > 0);
    const gravelLines = new Set(
      readFileSync(gravelFile, "utf-8").split("\n").filter((l) => l.length > 0)
    );

    let rockyouInGravel = 0;
    let gravelNotInRocks = 0;

    const rocksSet = new Set(rocksLines);
    for (const hash of gravelLines) {
      if (!rocksSet.has(hash)) gravelNotInRocks++;
      if (rockyouHashes.has(hash.toUpperCase())) rockyouInGravel++;
    }

    let unaccounted = 0;
    for (const hash of rocksLines) {
      if (!gravelLines.has(hash) && !rockyouHashes.has(hash.toUpperCase())) {
        unaccounted++;
      }
    }

    if (rockyouInGravel > 0 || gravelNotInRocks > 0 || unaccounted > 0) {
      console.log(
        `  ${batchName}: VIOLATION — rockyou_in_gravel=${rockyouInGravel}, ` +
          `gravel_not_in_rocks=${gravelNotInRocks}, unaccounted=${unaccounted}`
      );
      violations++;
    } else {
      verified++;
    }

    if ((verified + violations) % 100 === 0) {
      console.log(`  ... ${verified + violations}/${rocksBatches.length} checked`);
    }
  }

  console.log(
    `\nVerification: ${verified} OK, ${violations} violations out of ${rocksBatches.length} batches`
  );
}

// =============================================================================
// Status Display
// =============================================================================

function showStatus(): void {
  const progress = loadProgress();
  const rocksBatches = getRocksBatches();

  let gravelCount = 0;
  if (existsSync(GRAVEL_DIR)) {
    gravelCount = readdirSync(GRAVEL_DIR).filter(
      (f) => f.startsWith("batch-") && f.endsWith(".txt")
    ).length;
  }

  console.log("");
  console.log("GravelFilter Status");
  console.log("==========================");
  console.log(`Rockyou source:    ${progress.rockyouPath || "(not set)"}`);
  console.log(`Rockyou hashes:    ${progress.rockyouEntries.toLocaleString()}`);
  console.log(`ROCKS batches:     ${rocksBatches.length}`);
  console.log(`GRAVEL batches:    ${gravelCount}`);
  console.log(`Batches filtered:  ${progress.batchesProcessed}`);
  console.log(`ROCKS hashes:      ${progress.totalRocksHashes.toLocaleString()}`);
  console.log(`Rockyou filtered:  ${progress.totalRockyouFiltered.toLocaleString()}`);
  console.log(`GRAVEL hashes:     ${progress.totalGravelHashes.toLocaleString()}`);
  console.log(`Started:           ${progress.startedAt}`);
  console.log(`Last updated:      ${progress.lastUpdated}`);

  if (progress.batchesProcessed < rocksBatches.length) {
    const remaining = rocksBatches.length - progress.batchesProcessed;
    console.log(`\nPending: ${remaining} batches — resume with: bun Tools/GravelFilter.ts`);
  } else {
    console.log(`\nComplete! All ${rocksBatches.length} batches filtered.`);
  }
}

// =============================================================================
// CLI
// =============================================================================

if (import.meta.main) {
  const args = process.argv.slice(2);

  if (args[0] === "--help" || args[0] === "-h") {
    console.log(`
GravelFilter - ROCKS → GRAVEL Filter (1:1 batch correspondence)

Usage:
  bun Tools/GravelFilter.ts                   Filter all rocks → gravel (with resume)
  bun Tools/GravelFilter.ts --no-resume        Start fresh (re-filter all)
  bun Tools/GravelFilter.ts --verify            Verify rocks/gravel invariant
  bun Tools/GravelFilter.ts --status            Show progress
  bun Tools/GravelFilter.ts --rockyou PATH      Use alternate rockyou.txt

Pipeline: rocks/batch-NNNN.txt → [filter - SHA1(rockyou.txt)] → gravel/batch-NNNN.txt
Invariant: rocks[N] - SHA1(rockyou) = gravel[N]

Method:
  1. Reads rockyou.txt from disk (plain text)
  2. Computes SHA-1 hash of every entry in memory
  3. Filters each rocks batch against the hash Set
  4. Writes corresponding gravel batch

No pre-computed indexes. No trust in intermediate files.

Memory: ~3GB (14.3M SHA-1 hashes in Set)
  If OOM: bun --max-old-space-size=4096 Tools/GravelFilter.ts

Requires:
  - rocks/ directory populated by RocksExtractor.ts
  - rockyou.txt (default: ${DEFAULT_ROCKYOU})
`);
    process.exit(0);
  }

  // Parse --rockyou PATH
  let rockyouPath = DEFAULT_ROCKYOU;
  const rockyouIdx = args.indexOf("--rockyou");
  if (rockyouIdx !== -1 && args[rockyouIdx + 1]) {
    rockyouPath = resolve(args[rockyouIdx + 1]);
  }

  if (args[0] === "--status") {
    showStatus();
    process.exit(0);
  }

  if (args[0] === "--verify") {
    try {
      await verifyBatches(rockyouPath);
    } catch (e) {
      console.error(`Error: ${(e as Error).message}`);
      process.exit(1);
    }
    process.exit(0);
  }

  const resume = !args.includes("--no-resume");

  try {
    await filterAllBatches({ resume, rockyouPath });
  } catch (e) {
    console.error(`Error: ${(e as Error).message}`);
    console.error((e as Error).stack);
    process.exit(1);
  }
}
